streams/hash/others/hash_functions/tiger/tiger.cpp

```diff
@@ -0,0 +1,249 @@
+#define round(a,b,c,x,mul) \
+	c ^= x; \
+	a -= t1[(uint8_t)(c)] ^ \
+		t2[(uint8_t)((c) >> (2 * 8))] ^ \
+		t3[(uint8_t)((c) >> (4 * 8))] ^ \
+		t4[(uint8_t)((c) >> (6 * 8))] ; \
+	b += t4[(uint8_t)((c) >> (1 * 8))] ^ \
+		t3[(uint8_t)((c) >> (3 * 8))] ^ \
+		t2[(uint8_t)((c) >> (5 * 8))] ^ \
+		t1[(uint8_t)((c) >> (7 * 8))]; \
+	b *= mul;
+
+#else /* for IA32 */
+
+#define round(a,b,c,x,mul) \
+	c ^= x; \
+	a -= t1[(uint8_t)(c)] ^ \
+		t2[(uint8_t)(((uint32_t)(c)) >> (2 * 8))] ^ \
+		t3[(uint8_t)((c) >> (4 * 8))] ^ \
+		t4[(uint8_t)(((uint32_t)((c) >> (4 * 8))) >> (2 * 8))] ; \
+	b += t4[(uint8_t)(((uint32_t)(c)) >> (1 * 8))] ^ \
+		t3[(uint8_t)(((uint32_t)(c)) >> (3 * 8))] ^ \
+		t2[(uint8_t)(((uint32_t)((c) >> (4 * 8))) >> (1 * 8))] ^ \
+		t1[(uint8_t)(((uint32_t)((c) >> (4 * 8))) >> (3 * 8))]; \
+	b *= mul;
+#endif /* CPU_X64 */
+
+#define pass(a,b,c,mul) \
+	round(a,b,c,x0,mul) \
+	round(b,c,a,x1,mul) \
+	round(c,a,b,x2,mul) \
+	round(a,b,c,x3,mul) \
+	round(b,c,a,x4,mul) \
+	round(c,a,b,x5,mul) \
+	round(a,b,c,x6,mul) \
+	round(b,c,a,x7,mul)
+
+#define pass_rr(a,b,c,mul,ii,nr) \
+    if(8*(ii)+1 <= (nr)) \
+	round(a,b,c,x0,mul) \
+    if(8*(ii)+2 <= (nr)) \
+    round(b,c,a,x1,mul) \
+    if(8*(ii)+3 <= (nr)) \
+	round(c,a,b,x2,mul) \
+    if(8*(ii)+4 <= (nr)) \
+	round(a,b,c,x3,mul) \
+    if(8*(ii)+5 <= (nr)) \
+	round(b,c,a,x4,mul) \
+    if(8*(ii)+6 <= (nr)) \
+	round(c,a,b,x5,mul) \
+    if(8*(ii)+7 <= (nr)) \
+	round(a,b,c,x6,mul) \
+    if(8*(ii)+8 <= (nr)) \
+	round(b,c,a,x7,mul)
+
+#define key_schedule { \
+	x0 -= x7 ^ I64(0xA5A5A5A5A5A5A5A5); \
+	x1 ^= x0; \
+	x2 += x1; \
+	x3 -= x2 ^ ((~x1)<<19); \
+	x4 ^= x3; \
+	x5 += x4; \
+	x6 -= x5 ^ ((~x4)>>23); \
+	x7 ^= x6; \
+	x0 += x7; \
+	x1 -= x0 ^ ((~x7)<<19); \
+	x2 ^= x1; \
+	x3 += x2; \
+	x4 -= x3 ^ ((~x2)>>23); \
+	x5 ^= x4; \
+	x6 += x5; \
+	x7 -= x6 ^ I64(0x0123456789ABCDEF); \
+}
+
+/**
+ * The core transformation. Process a 512-bit block.
+ *
+ * @param state the algorithm state
+ * @param block the message block to process
+ */
+static void rhash_tiger_process_block(uint64_t state[3], uint64_t* block, unsigned nr)
+{
+    /* Optimized for GCC IA32.
+       The order of declarations is important for compiler. */
+    uint64_t a, b, c;
+    uint64_t x0, x1, x2, x3, x4, x5, x6, x7;
+#ifndef CPU_X64
+    uint64_t tmp;
+    char i;
+#endif
+
+    x0 = le2me_64(block[0]); x1 = le2me_64(block[1]);
+    x2 = le2me_64(block[2]); x3 = le2me_64(block[3]);
+    x4 = le2me_64(block[4]); x5 = le2me_64(block[5]);
+    x6 = le2me_64(block[6]); x7 = le2me_64(block[7]);
+
+    a = state[0];
+    b = state[1];
+    c = state[2];
+
+    /* passes and key shedules */
+#ifndef CPU_X64
+    for (i = 0; i < 3; i++) {
+        if (i != 0) key_schedule;
+        pass_rr(a, b, c, (i == 0 ? 5 : i == 1 ? 7 : 9), i, nr);
+        tmp = a;
+        a = c;
+        c = b;
+        b = tmp;
+    }
+#else
+    pass_rr(a, b, c, 5, 0, nr);
+	key_schedule;
+    pass_rr(c, a, b, 7, 0, nr);
```

Comment:
Mistake for !CPU_X64:

i=0: pass_rr(a, b, c, 5, 0, nr);
i=1: pass_rr(c, a, b, 7, **1**, nr);
i=2: pass_rr(b, c, a, 9, **2**, nr);

(It is just loop unroll here).

Reply:
good catch!